app/auth/web.py

```diff
@@ -80,140 +59,243 @@
+            time.sleep(3)
+            timeout -= 3
+            val = store.get(key)
+        if val:
+            store.delete(key)
+            return val
         else:
             logger.debug("Timeout while waiting for request {}".format(t))
             return '{"error": "timeout"}'
     else:
-        return "You can copy/paste the following tokens if needed and close this page: <br> Access Token: {}<br>Refresh Token: {}".format(
-            request.cookies.get('access_token'), request.cookies.get('refresh_token'))
+
+        if 'token' not in session:
+            return await app.make_response(redirect("{}?redirect_url={}".format(url_for('login'), quote_plus(url_for('info')))))
+
+        try:
+            a = jwt.decode(
+                session['token'],
+                app.config['OIDC_PUBLIC_KEY'],
+                algorithms=JWT_ALGORITHM,
+                audience=app.config['OIDC_CLIENT_ID']
+            )  # TODO: logout and redirect if fails because of expired
+
+            return "You can copy/paste the following tokens if needed and close this page: <br> Access Token: {}<br>Refresh Token: {}".format(
+                store.get(get_key_for_user(a, 'kc_access_token')).decode(), store.get(get_key_for_user(a, 'kc_refresh_token')).decode())
+
+        except jwt.ExpiredSignatureError:
+            return await app.make_response(redirect("{}?redirect_url={}".format(url_for('login'), quote_plus(url_for('info')))))
+
+
+
+@app.route(urljoin(app.config['SERVICE_PREFIX'], 'auth/user'))
+async def user():
+
+    if 'token' not in session:
+        return await app.make_response(redirect("{}?redirect_url={}".format(url_for('login'), quote_plus(url_for('user')))))
+    try:
+        a = jwt.decode(
+            session['token'],
+            app.config['OIDC_PUBLIC_KEY'],
+            algorithms=JWT_ALGORITHM,
+            audience=app.config['OIDC_CLIENT_ID']
+        )  # TODO: logout and redirect if fails because of expired
+
+        return store.get(get_key_for_user(a, 'kc_id_token')).decode()
+
+    except jwt.ExpiredSignatureError:
+            return await app.make_response(redirect("{}?redirect_url={}".format(url_for('login'), quote_plus(url_for('user')))))
 
 
 @app.route(urljoin(app.config['SERVICE_PREFIX'], 'auth/logout'))
 async def logout():
-    logout_url = app.config['OIDC_ISSUER'] + '/protocol/openid-connect/logout?redirect_uri=' + \
-                 request.args.get('redirect_url')
+
+    logout_url = '{}/protocol/openid-connect/logout?{}'.format(
+        app.config['OIDC_ISSUER'],
+        urllib.parse.urlencode({'redirect_uri': request.args.get('redirect_url')}),
+    )
     response = await app.make_response(redirect(logout_url))
-    response.delete_cookie('access_token', domain=COOKIE_DOMAIN)
-    response.delete_cookie('refresh_token', domain=COOKIE_DOMAIN)
-    response.delete_cookie('id_token', domain=COOKIE_DOMAIN)
+
+    a = jwt.decode(session['token'], verify=False)
```

Comment:
For me the logout is still failing here. The session does not contain a 'token', in fact it's entirely empty. However, if I check the content of the Redis store directly, the token is there. However, if I add
```
 import asyncio
 await asyncio.sleep(1)
```
at the beginning of the method, things work fine. So I assume that the Redis session lookup should be treated ("awaited") as an asynchronous call. Maybe we have to drop the convenience of the standard session implementation here and connect to Redis explicitly using something like `aioredis` (asynchronous Redis client library)?

Reply:
Can you also try the /api/auth/user or /api/auth/info (after a successful login)? Those two endpoints do also need to read the session to get the token. (if there was a general issue about accessing the session data, the login would have failed retrieving the state, I think it is just a matter of refactoring this endpoint to do the session access in the synchronous part of the method, i.e. before the first yield/await).

Reply:
Thanks @ebiiii for calling it a "trick" while it is more of a bug in the UI ;)